napari/utils/colormaps/colormap.py

```diff
@@ -460,48 +556,170 @@ def _zero_preserving_modulo_numpy(
-        n : int
-            The divisor.
-        dtype : np.dtype
-            The desired dtype for the output array.
-        to_zero : int, optional
-            A specific value to map to 0. (By default, 0 itself.)
 
-        Returns
-        -------
-        np.ndarray
-            The result: 0 for the ``to_zero`` value, ``values % n + 1``
-            everywhere else.
-        """
-        result = np.empty_like(values, dtype=dtype)
+def _cast_direct_labels_to_minimum_type_numpy(
+    data: np.ndarray, direct_colormap: DirectLabelColormap
+) -> np.ndarray:
+    """
+    Cast direct labels to the minimum type.
 
-        for i in numba.prange(values.size):
-            if values.flat[i] == to_zero:
-                result.flat[i] = 0
-            else:
-                result.flat[i] = (values.flat[i] - 1) % n + 1
+    Parameters
+    ----------
+    data : np.ndarray
+        The input data array.
+    direct_colormap : DirectLabelColormap
+        The direct colormap.
 
-        return result
+    Returns
+    -------
+    np.ndarray
+        The casted data array.
+    """
+    max_value = max(x for x in direct_colormap.color_dict if x is not None)
+    if max_value > 2**16:
+        raise RuntimeError(  # pragma: no cover
+            "Cannot use numpy implementation for large values of labels "
+            "direct colormap. Please install numba."
+        )
+    dtype = minimum_dtype_for_labels(direct_colormap.unique_colors_num() + 2)
+    label_mapping = direct_colormap.values_mapping_to_minimum_values_set()[0]
+
+    mapper = np.full((max_value + 2), DEFAULT_VALUE, dtype=dtype)
+    for key, val in label_mapping.items():
+        if key is None:
+            continue
+        mapper[key] = val
+
+    if data.dtype.itemsize > 2:
+        data = np.clip(data, 0, max_value + 1)
+    return mapper[data]
+
+
+def _zero_preserving_modulo_jit(
+    values: np.ndarray, n: int, dtype: np.dtype, to_zero: int = 0
+) -> np.ndarray:
+    """``(values - 1) % n + 1``, but with one specific value mapped to 0.
+
+    This ensures (1) an output value in [0, n] (inclusive), and (2) that
+    no nonzero values in the input are zero in the output, other than the
+    ``to_zero`` value.
+
+    Parameters
+    ----------
+    values : np.ndarray
+        The dividend of the modulo operator.
+    n : int
+        The divisor.
+    dtype : np.dtype
+        The desired dtype for the output array.
+    to_zero : int, optional
+        A specific value to map to 0. (By default, 0 itself.)
+
+    Returns
+    -------
+    np.ndarray
+        The result: 0 for the ``to_zero`` value, ``values % n + 1``
+        everywhere else.
+    """
+    result = np.empty_like(values, dtype=dtype)
+
+    for i in prange(values.size):
+        if values.flat[i] == to_zero:
+            result.flat[i] = 0
+        else:
+            result.flat[i] = (values.flat[i] - 1) % n + 1
+
+    return result
+
+
+def _cast_direct_labels_to_minimum_type_for_jit(
```

Comment:
I don't understand the function hierarchy here... Why does `_cast_direct_labels_to_minimum_type_for_jit` call `_cast_direct_labels_to_minimum_type_jit`? Something needs to get modified in these names to make it clear what each function does... I would also request a docstring like the one I did [in #6467](https://github.com/napari/napari/pull/6467/commits/785866fb4ce33a1221cb77c7e2e20c669e2b7353#diff-52091d3e98eca6f79e2e9847856adcb9eeda9cbb7eacff0f94883b467b6824f3R382-R395) that explains what happens to each data type — is there a copy, is there hashing or direct mapping, etc. Maybe what I am asking for is one level up from this one...